### PR TITLE
Python: Cleanup Flask models now that we have API graphs

### PR DESCRIPTION
**Pull request by RasmusWL**
Now that we have API graphs, we don't need to follow the old structure that type-tracking required. This is my take on how we can model things here in the glorious future. We might be able to make even more improvements in the future, but for now, I think this looks really nice :+1: -- let me know if there are things you don't agree with @yoff and @tausbn, I would suggest we take some time to talk this over in our meeting on Monday.

I think we're at a point where we can also allow access to these models from queries, which is also something I want.

### CODE REVIEW

File: python/ql/src/semmle/python/frameworks/Flask.qll [hunk 1]
```diff
@@ -15,127 +15,84 @@ private import semmle.python.ApiGraphs
...
+  module FlaskApp {
+    /** Gets a reference to the `flask.Flask` class. */
+    API::Node classRef() { result = API::moduleImport("flask").getMember("Flask") }
 
-      /** Gets a reference to the `add_url_rule` method on an instance of `flask.Blueprint`. */
-      API::Node add_url_rule() { result = instance_attr("add_url_rule") }
-    }
+    /** Gets a reference to an instance of `flask.Flask` (a flask application). */
+    API::Node instance() { result = classRef().getReturn() }
+  }
```
Comment (tausbn) [inline python/ql/src/semmle/python/frameworks/Flask.qll:62]: I feel like this could be encapsulated in a common pattern, say as follows:
```ql
// This should probably live in the API module
class Class extends Node {

  ... // Not really sure about the exact format of the charpred...

  /** Gets an instance of this class. */
  Node getInstance() { result = this.getReturn() }
}

// And then, in the model

class FlaskApp extends API::Class {
  FlaskApp() {this = API::moduleImport("flask").getMember("Flask") }
}
// And then, when used:
predicate foo() {
  blah = any(FlaskApp fa) // for the class
  and
  blah = any(FlaskApp fa).getInstance() // for instances
}
```
With this, we could then also capture the `[FlaskApp::classRef(), FlaskApp::instance()]` pattern as a method of `API::Class` (though I don't know exactly what would be a good name for it).

File: python/ql/src/semmle/python/frameworks/Flask.qll [hunk 2]
```diff
@@ -315,8 +322,14 @@ module Flask {
...
    * See https://flask.palletsprojects.com/en/1.1.x/api/#flask.Request
    */
-  private class RequestSource extends RemoteFlowSource::Range {
-    RequestSource() { this = request().getAUse() }
+  private class FlaskRequestSource extends RemoteFlowSource::Range {
+    FlaskRequestSource() {
+      this = request().getAUse() and
+      not any(Import imp).contains(this.asExpr()) and
+      not exists(ControlFlowNode def | this.asVar().getSourceVariable().hasDefiningNode(def) |
+        any(Import imp).contains(def.getNode())
```
Comment (tausbn) [inline python/ql/src/semmle/python/frameworks/Flask.qll:330]: This one raised a few eyebrows for me. Do we perhaps need this pattern encapsulated in a utility method on `API::Node`?

**Review by tausbn — CHANGES_REQUESTED** [review]
Overall I think this looks great!

File: python/ql/src/semmle/python/frameworks/Flask.qll (hunk 2 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Flask.qll:330]: I agree it would be nice to have an easy way to filter out "usage" in imports :+1:

File: python/ql/src/semmle/python/frameworks/Flask.qll (hunk 1 above)
Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Flask.qll:62]: I think being able to capture class-attributes with something more elegant than `[FlaskApp::classRef(), FlaskApp::instance()].getMember("class_attribute_name")` would be nice long term.

However, the fact that `instance()` have a `API::Node` result instead of `DataFlow::Node` goes against my argument in https://github.com/github/codeql/pull/5208/commits/35876f1939dd1ad38ef84870b8fd2063cc80813a :grimacing: (but it really needs `API::Node` to be useful)

So I guess I would like to hold off with this until we're able to use it everywhere. When we have the "def" side of API graphs, it could allow us back in the very familiar QL modeling territory.

Allowing myself to experiment with this idea a bit, with this setup, we could model the "tricky" response class as

```codeql
class FlaskResponse extends API::Class {
  FlaskResponse() {
    this = API::moduleImport("flask").getMember("Response")
    or
    this = any(FlaskApp).getClassMember("response_class")
  }

  override API::Node getInstance() { 
    result = super.getInstance()
    or
    result = API::moduleImport("flask").getMember("make_response").getACall()
    or
    ...
  }
```

However, for some classes, I think we want to model instances as an open union. 

If for example `my_flask_library` also contains functions that can create instances `flask.Response`, with the setup above, we would need to edit the `FlaskResponseClass` implementation in `Flask.qll`, and I would rather not force us into always doing this.

So we could make `FlaskResponseClass` abstract, but then what do we specify in the char-pred for the instance that comes from `my_flask_library.make_response`?

This leads me to think we need a separate modeling of classes and class instances. Possibly something like

```codeql
class FlaskResponseClass extends API::Class {
  FlaskResponseClass() {
    this = "flask.Response"
  }

  override API::Node getARef() {
    this = API::moduleImport("flask").getMember("Response")
    or
    this = any(FlaskApp).getClassMember("response_class")
  }
}

abstract class FlaskResponseInstance extends API::ClassInstance {
  override API::Class getClass() { result instanceof FlaskResponseClass }
}

class FlaskResponseDirectRef extends FlaskResponseInstance {
  FlaskResponseDirectRef() { this = any(FlaskResponseClass cls).getARef().getReturn() }
}

class FlaskMakeResponseCall extends FlaskResponseInstance {
  FlaskMakeResponseCall() { this = API::moduleImport("flask").getMember("make_response").getACall() }
}
```

But again, the modeling of references to a class is not quite open enough, since there is no extension-point to allow `my_flask_library.Response` to be a reference to that class.

**In conclusion**: Great idea, from my point of view, this seems tricky to get right. I would rather have us think this through a bit more, and not implement anything right now.

**Comment by RasmusWL** [comment]
I'm just thought about how it might be a bit strange that we are exposing both `Flask::Views::View` and `Flask::FlaskViewClass`. The way I'm thinking about framework models is that there is part of the library that is interesting to model (in this case, the `flask.views.View` class), and that there are uses of the library that is interesting to model (in this case, classes that use `flask.views.View` as a subclass). 

For this specific example, I'm thinking that we could actually just remove `Flask::Views::View`, since it's probably not that interesting in itself, and only import for defining `Flask::FlaskViewClass` in the current setup. The only argument for keeping `Flask::Views::View` is to make the `subclassRef` be based on an open-union of some sort (thereby enabling other framework modeling).

Tricky stuff :thinking:

File: python/ql/src/semmle/python/frameworks/Flask.qll [hunk 3]
```diff
@@ -15,127 +15,84 @@ private import semmle.python.ApiGraphs
...
-       */
-      API::Node response_class() { result = [classRef(), instance()].getMember("response_class") }
+    module View {
+      /** Gets a reference to the `flask.views.View` class or any subclass. */
+      API::Node subclassRef() {
+        result =
+          API::moduleImport("flask")
+              .getMember("views")
+              .getMember(["View", "MethodView"])
+              .getASubclass*()
```
Comment (yoff) [inline python/ql/src/semmle/python/frameworks/Flask.qll:33]: Why is `"MethodView"` included here? Should the QLDoc be updated?

Comment (RasmusWL) [reply python/ql/src/semmle/python/frameworks/Flask.qll]: fixed :+1:

**Comment by RasmusWL** [comment]
Updated with the minor changes we discussed earlier today, so ready for a final review :+1:

**Comment by RasmusWL** [comment]
I'm more in the mood for dark-shipping this for now, so adding the label to avoid writing a change note. I would rather do a public splash about this in the future, when we have at least ported all the framework modeling away from the old very verbose type tracking code.

**Review by tausbn — APPROVED** [review]
I think there's still room for improvement, but I agree that this is best kept until we have built out the modelling capabilities of API graphs a bit. 
I'm happy to see this merged in its current state.

**Review by yoff — APPROVED** [review]
LGTM

**Comment by yoff** [comment]
> I think there's still room for improvement, but I agree that this is best kept until we have built out the modelling capabilities of API graphs a bit.
> I'm happy to see this merged in its current state.

Indeed, but I am happy to consider it a "gold-standard-for-now" for library modelling. When we document it, we will have to also comment on the approach. I imagine it is similar to API graphs, in that one should pull out all interesting classes and methods in addition to all interesting concepts.
